server/etcdserver/txn/txn_test.go

```diff
@@ -37,231 +37,236 @@ import (
-				Success: []*pb.RequestOp{
-					{
-						Request: &pb.RequestOp_RequestPut{
-							RequestPut: &pb.PutRequest{
-								Lease: 123,
-							},
-						},
-					},
+		expectError: "etcdserver: key not found",
+	},
+	{
+		name: "Put with ignore lease without previous key should fail",
+		op: &pb.RequestOp{
+			Request: &pb.RequestOp_RequestPut{
+				RequestPut: &pb.PutRequest{
+					IgnoreLease: true,
 				},
 			},
-			expectError: "lease not found",
 		},
-		{
-			name:       "Put with existing lease should succeed",
-			setupLease: 123,
-			txn: &pb.TxnRequest{
-				Success: []*pb.RequestOp{
-					{
-						Request: &pb.RequestOp_RequestPut{
-							RequestPut: &pb.PutRequest{
-								Lease: 123,
-							},
-						},
-					},
+		expectError: "etcdserver: key not found",
+	},
+	{
+		name:  "Put with ignore value with previous key should succeeded",
+		setup: testSetup{key: []byte("ignore-value")},
+		op: &pb.RequestOp{
+			Request: &pb.RequestOp_RequestPut{
+				RequestPut: &pb.PutRequest{
+					IgnoreValue: true,
+					Key:         []byte("ignore-value"),
 				},
 			},
 		},
-		{
-			name: "Put with ignore value without previous key should fail",
-			txn: &pb.TxnRequest{
-				Success: []*pb.RequestOp{
-					{
-						Request: &pb.RequestOp_RequestPut{
-							RequestPut: &pb.PutRequest{
-								IgnoreValue: true,
-							},
-						},
-					},
+	},
+	{
+		name:  "Put with ignore lease with previous key should succeed ",
+		setup: testSetup{key: []byte("ignore-lease")},
+		op: &pb.RequestOp{
+			Request: &pb.RequestOp_RequestPut{
+				RequestPut: &pb.PutRequest{
+					IgnoreLease: true,
+					Key:         []byte("ignore-lease"),
 				},
 			},
-			expectError: "etcdserver: key not found",
 		},
-		{
-			name: "Put with ignore lease without previous key should fail",
+	},
+}
+
+func TestCheckTxn(t *testing.T) {
+	type txnTestCase struct {
+		name        string
+		setup       testSetup
+		txn         *pb.TxnRequest
+		expectError string
+	}
+	testCases := []txnTestCase{}
+	for _, tc := range append(rangeTestCases, putTestCases...) {
+		testCases = append(testCases, txnTestCase{
+			name:  tc.name,
+			setup: tc.setup,
 			txn: &pb.TxnRequest{
 				Success: []*pb.RequestOp{
-					{
-						Request: &pb.RequestOp_RequestPut{
-							RequestPut: &pb.PutRequest{
-								IgnoreLease: true,
-							},
-						},
-					},
+					tc.op,
 				},
 			},
-			expectError: "etcdserver: key not found",
-		},
-		{
-			name:     "Put with ignore value with previous key should succeeded",
-			setupKey: []byte("ignore-value"),
-			txn: &pb.TxnRequest{
-				Success: []*pb.RequestOp{
-					{
-						Request: &pb.RequestOp_RequestPut{
-							RequestPut: &pb.PutRequest{
-								IgnoreValue: true,
-								Key:         []byte("ignore-value"),
-							},
-						},
-					},
-				},
+			expectError: tc.expectError,
+		})
+	}
+	testCases = append(testCases, txnTestCase{
+		name:  "Invalid range on failed path should succeed",
+		setup: invalidRangeTestCase.setup,
```

Comment:
pls make this case independent. In other words, pls do not cross link to/depend on other case. Also `invalidRangeTestCase.setup` isn't assigned a value at all.

Reply:
DOne